sdk/eventplugin/eventplugin.proto

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+syntax = "proto3";
+
+package eventplugin.v1;
+
+import "google/protobuf/struct.proto";
+import "sdk/plugin/pb/backend.proto";
+import "sdk/logical/event.proto";
+
+option go_package = "github.com/hashicorp/vault/sdk/eventplugin";
+
+message InitializeRequest {}
+message InitializeResponse {}
+
+message SubscriptionEvent {
+	// only a single subscription ID will be used in a stream, so the value MAY be omitted in subsequent messages,
+	// but is REQUIRED for the first message. This is included for future-proofing having a plugin handle multiple
+	// subscriptions per plugin instance.
+	string subscription_id = 1;
+	string eventJson = 2;
+}
+
+message SubscribeRequest {
+	// parameters to set up the request, e.g., queue name, API key
+	google.protobuf.Struct config = 1;
+	// globally unique subscription ID, generated by Vault
+	string subscription_id = 2;
+	// whether to verify the connection immediately.
+	bool verify_connection = 3;
+}
+
+message UnsubscribeRequest {
+	string subscription_id = 1;
+}
+
+message UnsubscribeResponse {}
+
+message SubscribeResponse {
+	pb.ProtoError err = 1;
+	// If true, Vault can retry sending events with this subscription on error.
+	// If false, the subscription is canceled on error.
+	bool retry = 2;
+}
+
+message SendSubscriptionEventsResponse {}
+
+message TypeRequest {}
+
+message TypeResponse {
+	// plugin type, e.g., "aws-sqs"
+	string plugin_type = 1;
+	// plugin version, e.g., "v1.0.0"
+	string plugin_version = 2;
+}
+
+message CloseRequest {}
+message CloseResponse {}
+
+// plugins should implement the server
+service EventSubscribePluginService {
+	rpc Initialize(InitializeRequest) returns (InitializeResponse);
+	// Start a new subscription.
+	rpc Subscribe(SubscribeRequest) returns (SubscribeResponse);
+	// Used by Vault to send events to a subscription. Only one stream is allowed per subscription.
```

Comment:
Trying to get my head around the model here - do we envisage mounting events plugins like other plugins, or is there only ever going to be a single instance of them and they'll handle multiple subscriptions?

Then there's a 1-1 mapping between subscription and stream - is there any case where a stream will restart or is closing a stream tantamount to calling Unsubscribe?

And there's going to be a many:many mapping from event filters to subscriptions that's not defined in this PR right?

Reply:
For now, there's a 1-1 mapping between subscriptions and plugin instances, but I wanted to leave the interface flexible about that, so that it could potentially support multiplexing natively without having to be retrofitted later.

I think closing the stream and unsubscribing should be synonymous.

And yes, a subscription will allow (potentially many) event type globs + event filters mapped to a subscription.